src/websocket_endpoint.cpp

```diff
@@ -108,5 +138,108 @@ namespace sensr
     }     
   }
 
+  /// Verify that one of the subject alternative names matches the given hostname
+  bool WebSocketEndPoint::VerifySubjectAlternativeName(const char * hostname, X509 * cert) {
+    STACK_OF(GENERAL_NAME) * san_names = NULL;    
+    san_names = (STACK_OF(GENERAL_NAME) *) X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL);
+    if (san_names == NULL) {
+        return false;
+    }    
+    int san_names_count = sk_GENERAL_NAME_num(san_names);    
+    bool result = false;    
+    for (int i = 0; i < san_names_count; i++) {
+        const GENERAL_NAME * current_name = sk_GENERAL_NAME_value(san_names, i);        
+        if (current_name->type != GEN_DNS) {
+            continue;
+        }        
+        char const * dns_name = (char const *) ASN1_STRING_get0_data(current_name->d.dNSName);        
+        // Make sure there isn't an embedded NUL character in the DNS name
+        if (ASN1_STRING_length(current_name->d.dNSName) != static_cast<int>(strlen(dns_name))) {
+            break;
+        }
+        // Compare expected hostname with the CN
+        result = (strcasecmp(hostname, dns_name) == 0);
+    }
+    sk_GENERAL_NAME_pop_free(san_names, GENERAL_NAME_free);    
+    return result;
+  }
+
+  /// Verify that the certificate common name matches the given hostname
+  bool WebSocketEndPoint::VerifyCommonName(char const * hostname, X509 * cert) {
+    // Find the position of the CN field in the Subject field of the certificate
+    int common_name_loc = X509_NAME_get_index_by_NID(X509_get_subject_name(cert), NID_commonName, -1);
+    if (common_name_loc < 0) {
+        return false;
+    }    
+    // Extract the CN field
+    X509_NAME_ENTRY * common_name_entry = X509_NAME_get_entry(X509_get_subject_name(cert), common_name_loc);
+    if (common_name_entry == NULL) {
+        return false;
+    }    
+    // Convert the CN field to a C string
+    ASN1_STRING * common_name_asn1 = X509_NAME_ENTRY_get_data(common_name_entry);
+    if (common_name_asn1 == NULL) {
+        return false;
+    }    
+    char const * common_name_str = (char const *) ASN1_STRING_get0_data(common_name_asn1);    
```

Comment:
Can we use static_cast?

Reply:
sorry compiler complain invalid type conversion.
I think we need to use reinterpret_cast